File: 1472-design-browser-history/1472-design-browser-history.py
class Node:
    def __init__(self,val):
        self.val = val
        self.next = None
        self.pre = None

class BrowserHistory:

    def __init__(self, homepage: str):
        self.curr = Node(homepage)
        
    def visit(self, url: str) -> None:
        self.curr.next = Node(url)
        self.curr.next.pre = self.curr
        self.curr = self.curr.next
        return

    def back(self, steps: int) -> str:
        for _ in range(steps):
            if self.curr.pre != None:
                self.curr = self.curr.pre
        return self.curr.val

    def forward(self, steps: int) -> str:
        for _ in range(steps):
            if self.curr.next != None:
                self.curr = self.curr.next
        return self.curr.val


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)

File: 1472-design-browser-history/1472-design-browser-history.java
class Node{
    String val;
    Node next;
    Node pre;
    public Node(String val){
        this.val = val;
    }
}

class BrowserHistory {
    Node curr;
    public BrowserHistory(String homepage) {
        curr = new Node(homepage);
    }
    
    public void visit(String url) {
        curr.next = new Node(url);
        curr.next.pre = curr;
        curr = curr.next;
        return ;
    }
    
    public String back(int steps) {
        for (int i = 0 ; i < steps ; i ++){
            if (curr.pre != null)
                curr = curr.pre;
        }
        return curr.val;
    }
    
    public String forward(int steps) {
        for (int i = 0 ; i < steps ; i ++){
            if (curr.next != null)
                curr = curr.next;
        }
        return curr.val;     
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */